source/code/plugins/agent_telemetry_script.rb

```diff
@@ -0,0 +1,371 @@
+require 'optparse'
+
+module OMS
+
+  require_relative 'agent_topology_request_script'
+
+  # Operation Types
+  SEND_BATCH = "SendBatch"
+
+  class AgentResourceUsage < StrongTypedClass
+    strongtyped_accessor :OMSMaxMemory, Integer
+    strongtyped_accessor :OMSMaxPercentMemory, Integer
+    strongtyped_accessor :OMSMaxUserTime, Integer
+    strongtyped_accessor :OMSMaxSystemTime, Integer
+    strongtyped_accessor :OMSAvgMemory, Integer
+    strongtyped_accessor :OMSAvgPercentMemory, Integer
+    strongtyped_accessor :OMSAvgUserTime, Integer
+    strongtyped_accessor :OMSAvgSystemTime, Integer
+    strongtyped_accessor :OMIMaxMemory, Integer
+    strongtyped_accessor :OMIMaxPercentMemory, Integer
+    strongtyped_accessor :OMIMaxUserTime, Integer
+    strongtyped_accessor :OMIMaxSystemTime, Integer
+    strongtyped_accessor :OMIAvgMemory, Integer
+    strongtyped_accessor :OMIAvgPercentMemory, Integer
+    strongtyped_accessor :OMIAvgUserTime, Integer
+    strongtyped_accessor :OMIAvgSystemTime, Integer
+  end
+  
+  class AgentQoS < StrongTypedClass
+    strongtyped_accessor :Operation, String
+    strongtyped_accessor :OperationSuccess, String
+    strongtyped_accessor :Message, String
+    strongtyped_accessor :Source, String
+    strongtyped_accessor :BatchCount, Integer
+    strongtyped_accessor :MinBatchEventCount, Integer
+    strongtyped_accessor :MaxBatchEventCount, Integer
+    strongtyped_accessor :AvgBatchEventCount, Integer
+    strongtyped_accessor :MinBatchSize, Integer
+    strongtyped_accessor :MaxBatchSize, Integer
+    strongtyped_accessor :AvgBatchSize, Integer
+    strongtyped_accessor :MinLocalLatency, Integer
+    strongtyped_accessor :MaxLocalLatency, Integer
+    strongtyped_accessor :AvgLocalLatency, Integer
+    strongtyped_accessor :NetworkLatency, Integer
+  end
+
+  class AgentTelemetry < StrongTypedClass
+    strongtyped_accessor :OSType, String
+    strongtyped_accessor :OSDistro, String
+    strongtyped_accessor :OSVersion, String
+    strongtyped_accessor :Region, String
+    strongtyped_accessor :ConfigMgrEnabled, String
+    strongtyped_accessor :ResourceUsage, AgentResourceUsage
+    strongtyped_accessor :QoS, Array # of AgentQoS
+  end
+  
+  class Telemetry
+
+    require 'json'
+    # require 'objspace'
```

Comment:
for now, still have issue with `kernel_require.rb:55:in `require': cannot load such file -- objspace (LoadError)` even after enabling the ObjectSpace extension during ruby build.

Reply:
Did you include the new ruby content in ruby.data after removing objspace in configure ?
because I can see ${{RUBY_DEST}}/lib/ruby/2.4.0/${{RUBY_ARCM}}/etc.so;
You should see this path: ${{RUBY_DEST}}/lib/ruby/2.4.0/${{RUBY_ARCM}}/objspace.so;

Will be installed here: /opt/microsoft/omsagent/ruby/lib/ruby/2.4.0/x86_64-linux/objspace.so

Reply:
🙏 thank you 🙏